### PR TITLE
Fix rule plugins

### LINKED ISSUE #46: Update rule plugins to use new allow/deny methods of base class
When I updated `rule_base`, I only implemented the changes for `registrationrule_disposableemails` to demonstrate how the changes should be used. I also added lang strings for fallback points etc.

Now we need to update the other rules to use the new `allow()` and `deny()` methods in the same way.

For usage see...

https://github.com/daledavies/moodle-tool_registrationrules/blob/main/rules/disposableemails/classes/rule.php#L80
https://github.com/daledavies/moodle-tool_registrationrules/blob/main/rules/disposableemails/classes/rule.php#L90
https://github.com/daledavies/moodle-tool_registrationrules/blob/main/rules/disposableemails/classes/rule.php#L97

### PR DESCRIPTION
**Pull request by daledavies**
- Moves creation of rule instance objects from `rule_checker` into `rule_instances_controller`.
- Renames `configurable` interface to `instance_configurable` and introduces a new `plugin_configurable` interface, allowing plugins to implement their own "global" settings.
- Fixes a few issues and errors found in some rule plugins.
- Addresses issue #46.

I'd probably recommend reviewing this commit by commit, rather than all at once.

### CODE REVIEW

File: classes/local/rule_instances_controller.php [hunk 1]
```diff
@@ -168,18 +168,72 @@ public function get_rule_instance_records(): array {
...
+        $activeinstances = [];
+        foreach ($this->get_rule_instances() as $instance) {
+            $instanceconfig = $instance->get_config();
+            if (!$instanceconfig->pluginenabled || !$instanceconfig->enabled) {
+                continue;
+            }
+            // If this rule plugin class implements plugin_configurable then we can cheeck if the
+            // requierd configuration options have been satisfied.
+            if (is_subclass_of($instance, 'tool_registrationrules\local\rule\plugin_configurable')) {
+                $pluginconfigured = call_user_func(
```
Comment (daledavies) [inline classes/local/rule_instances_controller.php]: I'm actually unsure if it would be better to just use a `method_exists` check and call the function using a variable class name if we don't have an instance already. Moodle docs suggest we should avoid `call_user_func` and I wonder if my use of `is_subclass_of` might be less maintainable.

So this would become something like the following, note this is untested...

```
$classname = 'registrationrule_' . $instance->get_type() . '\rule';
if (method_exists($classname, 'is_plugin_configured')) {
    if (!$classname::is_plugin_configured()) {
        continue;
    }
}

Comment (daledavies) [reply classes/local/rule_instances_controller.php]: I've gone with a slightly different approach, leaving the `is_subclass_of` check in place, but removing `call_user_func`. This change is made in a new commit.... https://github.com/daledavies/moodle-tool_registrationrules/pull/55/commits/de7e1a79d8f4506a55b494a63915b23bcc419904

File: classes/local/rule_instances_controller.php [hunk 2]
```diff
@@ -168,18 +168,72 @@ public function get_rule_instance_records(): array {
...
+            $ruleinstance = new $pluginrule($instance);
+            if (!$ruleinstance instanceof rule\rule_base) {
+                debugging("Rule $pluginrule does not extend rule_base", DEBUG_DEVELOPER);
+                continue;
+            }
+            $instances[$ruleinstance->get_id()] = $ruleinstance;
+        }
+
+        return $instances;
+    }
```
Comment (phager-at) [inline classes/local/rule_instances_controller.php]: I'd like to propose we stick to only checking for the defined interfaces. Do we really care, if it's not a rule\rule_base subclass if it implements all interface methods we need and expect?

Also we can change return type to `tool_registrationrules\local\rule\rule_interface[]` here.

File: classes/local/rule_instances_controller.php [hunk 3]
```diff
@@ -168,18 +168,72 @@ public function get_rule_instance_records(): array {
...
+     * Return a rule instance object matching the given instanceid.
      *
      * @param int $instanceid
-     * @return \stdClass|null A single rule instance record.
+     * @return rule\rule_base|null A single rule instance record.
      */
-    public function get_rule_instance_by_id(int $instanceid) {
-        $instance = array_column($this->ruleinstances, null, 'id')[$instanceid];
-        if (!$instance) {
+    public function get_rule_instance_by_id(int $instanceid): rule\rule_base {
```
Comment (phager-at) [inline classes/local/rule_instances_controller.php]: As above, we should return a `tool_registrationrules\local\rule\rule_interface` here, as that's what the interface is for.

File: classes/local/rule_settings.php [hunk 4]
```diff
@@ -49,14 +49,8 @@ public static function from_rule_instance(int $instanceid): static {
...
-        $extrafields = json_decode($instancerecord->other);
-
-        foreach ($extrafields as $fieldname => $value) {
-            $instancerecord->$fieldname = $value;
-        }
-
-        $form = new static($instancerecord->type, $instanceid);
-        $form->set_data($instancerecord);
+        $form = new static($instancerecord->get_type(), $instanceid);
+        $form->set_data($instancerecord->get_config());
```
Comment (phager-at) [inline classes/local/rule_settings.php]: Currently we have a proper rule instance object here, so we could rename the variable to $ruleinstance or similar to be more precise.

File: rules/hcaptcha/classes/rule.php [hunk 5]
```diff
@@ -99,11 +100,11 @@ public function post_data_check(array $data): ?rule_check_result {
             );
         }
 
-        // If empty or false the captcha failed and the result is negative.
-        if (!empty($response->success)) {
+        // The captcha failed so deny registration.
+        if (!$response->success) {
```
Comment (phager-at) [inline rules/hcaptcha/classes/rule.php]: We should not access `$response`'s properties directly, as `json_decode()` returns `null` if it can't be decoded…

So we should check for a correctly decoded object, before accessing the data.

File: rules/hcaptcha/settings.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    $settings->add($setting);
+}
+
+$settings->add(new admin_setting_configtext('registrationrule_hcaptcha/hcaptcha_sitekey',
+    get_string('sitekey', 'registrationrule_hcaptcha'),
+    null, null, PARAM_ALPHANUMEXT));
+
+$settings->add(new admin_setting_configtext('registrationrule_hcaptcha/hcaptcha_secret',
+get_string('secret', 'registrationrule_hcaptcha'),
+    null, null, PARAM_ALPHANUMEXT));
```
Comment (phager-at) [inline rules/hcaptcha/settings.php]: Should this be at least a `admin_setting_configpasswordunmask`? It's still stored in plaintext, but it isn't visible right away in UI…

File: classes/local/rule/rule_base.php [hunk 7]
```diff
@@ -51,9 +57,38 @@ abstract class rule_base implements rule_interface {
      */
     public function __construct(stdClass $config) {
         $this->config = $config;
+        $this->id = $this->config->id;
+        $this->type = $this->config->type;
         $this->result = new rule_check_result();
     }
```
Comment (phager-at) [inline classes/local/rule/rule_base.php]: I'm a little concerned how the constructor parameter $config changed over time.
It seems not clear, what we expect here, without reading the code.
In the beginning we simply needed the (already decoded) configuration object (=extra rule instance-specific configuration), but now we're expecting a mixture of database record ID, plugin type (which is also recorded that way in DB) and the already decoded specific configuration (in DB = JSON; expected here: ready to use object properties?)…

We may have to clean that up using a consistent and clearly communicated approach.

**Review by phager-at — COMMENTED** [review]
TL;DR:
- +1 for removing methods from base class, which are only needed in some plugins…
- we should follow the pattern of using interfaces for our expected available methods in some so we can also define our `rule_interface[]` as return value for get_rule_instances()
- we may be able to improve our rule instances' constructor  and expected attribute data and/or communicate them more clearly
- admin_setting_configpasswordunmask instead of plain text for secret?

File: classes/local/rule_instances_controller.php (hunk 2 above)
Comment (daledavies) [reply classes/local/rule_instances_controller.php]: The counter argument is whether we need a `rule_interface` at all? Are we more interested in whether it is _actually_ a duck, or if it just quacks like a duck? :)

I'm not too bothered either way, but currently we have an interface that doesn't define all the methods of `rule_base`, so if we just check for the interface then we probably need to improve the interface to include all expected properties.

File: classes/local/rule_settings.php (hunk 4 above)
Comment (daledavies) [reply classes/local/rule_settings.php]: Ahhh this was on my mind to change and I forgot all about it! Thanks for the reminder :)

File: rules/hcaptcha/classes/rule.php (hunk 5 above)
Comment (daledavies) [reply rules/hcaptcha/classes/rule.php]: Not sure why this has the outdated label, but yes you are correct :)

File: rules/hcaptcha/settings.php (hunk 6 above)
Comment (daledavies) [reply rules/hcaptcha/settings.php]: Makes sense :)

File: classes/local/rule/rule_base.php (hunk 7 above)
Comment (daledavies) [reply classes/local/rule/rule_base.php]: Yeah on reflection I think first the $config param is badly named, so this needs to be changed at the very least.

Maybe we should be constructing a rule object with the correct properties so it doesn't have to transform them itself internally then?

But something like this maybe doesn't make sense as we may as well just pass it the DB record...

$id, $type, $points, $fallbackpoints, $sortorder, $config

So maybe we should (as per the comment on the `rule_base` constructor), construct the object with the DB record and let it work out what is "config" or not.

File: classes/local/rule_instances_controller.php (hunk 2 above)
Comment (phager-at) [reply classes/local/rule_instances_controller.php]: Big plus for the  interface would be: it's clearly defined what we expect and in which way exactly. If we just query for the methods, we just know if we can call them at all. But we don't know if they consume our expected parameters and provide expected return values (and types).

File: classes/local/rule/rule_base.php (hunk 7 above)
Comment (daledavies) [reply classes/local/rule/rule_base.php]: I have an idea, I will be back shortly with some more code

Comment (daledavies) [reply classes/local/rule/rule_base.php]: Roughly, we'll construct the rule object with params as follows ($id, $type, $points, $fallbackpoints, $sortorder, $otherconfig). Maybe the constructor can validate `$otherconfig` against what's returned from `get_instance_settings_fields()`.

File: classes/local/rule_instances_controller.php (hunk 2 above)
Comment (daledavies) [reply classes/local/rule_instances_controller.php]: I've implemented this and I will push as a new commit in the morning for you to see, but it has left me confused about whether this is a good idea.

The PHP docs state...


> Although they are supported, including [constructors](https://www.php.net/manual/en/language.oop5.decon.php#language.oop5.decon.constructor) in interfaces is strongly discouraged. Doing so significantly reduces the flexibility of the object implementing the interface. Additionally, constructors are not enforced by inheritance rules, which can cause inconsistent and unexpected behavior.

We kind of enforce that plugin classes are instantiated with particular parameters and the only way for a developer to know this would be to extend from our abstract base class, unless we include the constructor in the interface, but this is strongly discouraged and wouldn't be enforced by PHP's inheritance rules anyway.

We could just not include the constructor in the interface, but I can't see why anyone would _not_ want to use our abstract class either.

File: classes/local/rule/rule_base.php (hunk 7 above)
Comment (daledavies) [reply classes/local/rule/rule_base.php]: OK so I did this and although there is more code it seems a lot simpler/easier to follow, plus I've thrown a few extra bits of validation in there too. I'll push it in the morning as a new commit.

File: classes/local/rule_instances_controller.php (hunk 2 above)
Comment (daledavies) [reply classes/local/rule_instances_controller.php]: In other words, the constructor is an implementation detail and shouldn't be part of an interface, if we need to enforce an implementation detail then we should at least use an abstract class.

I'm actually thinking we should probably enforce the use of our abstract class and maybe even do away with the rule_interface entirely as it seems superfluous. The instances controller needs a rule plugin class to have certain implementation details (i.e. the constructor signature), not just present certain methods. 

We could do this but also create separate interfaces for pre/post data functions and extending the sign-up form. That way we have a proper (not abstract) base class and a range of optional interfaces that rule plugin classes could choose to implement. We would check for those before calling the methods they contain, and if we check the instance started as our base class then we know it will do all the things we want it to do.

Of course, there are conflicting opinions online about this so I'm unable to form a consensus. I'm interested in hearing your thoughts 😁

Comment (daledavies) [reply classes/local/rule_instances_controller.php]: The more I think about only checking for `rule_interface`, the more I'm inclined to say this isn't a good idea. The constructor signature is an implementation detail (which should not be what interfaces are for), and as we require this to be enforced then we should be enforcing the use of at least an abstract base class by sub-classes. 

Also in this case an additional interface is superfluous as it would only ever be used by the base class and would also define exactly the same methods as the base class.

This has given me the opportunity to rethink what we're doing with `extend_form()`, `pre_data_check()` and `post_data_check()` as this could definitely be improved. I think we should create separate interfaces for each feature and then check to see if an interface has been implemented by the sub-plugin before calling the respective methods, rather than forcing subplugins to define the methods but just return null if they are not required.

File: rules/hcaptcha/classes/rule.php (hunk 5 above)
Comment (daledavies) [reply rules/hcaptcha/classes/rule.php]: Done

File: classes/local/rule_settings.php (hunk 4 above)
Comment (daledavies) [reply classes/local/rule_settings.php]: Done

File: rules/hcaptcha/settings.php (hunk 6 above)
Comment (daledavies) [reply rules/hcaptcha/settings.php]: Done

File: classes/local/rule/rule_base.php (hunk 7 above)
Comment (daledavies) [reply classes/local/rule/rule_base.php]: Done.

As mentioned I've completely reworked this in a new commit... `Clean up definition of config and sub-plugin class properties`

File: classes/local/rule_instances_controller.php (hunk 2 above)
Comment (daledavies) [reply classes/local/rule_instances_controller.php]: As discussed above I've gone ahead and removed the `rule_interface`, moving abstract rule_base methods to optional interfaces.

The result is in a new commit... `Move abstract rule_base methods to optional interfaces`

Although it is not what you proposed, I do think this makes for a much cleaner and more consistent approach.

**Comment by daledavies** [comment]
@phager-at Just need to fix the failing checks and it'll be back over to you.

**Comment by daledavies** [comment]
All done 😃

File: classes/local/rule/rule_base.php (hunk 7 above)
Comment (phager-at) [reply classes/local/rule/rule_base.php]: I like the split parameters. What I'm just thinking about is: we currently need the instance configuration, even if there is none. Also our DB column `other` is nullable, but we're assuming everywhere, that there is configuration.

It should at least be optional. Maybe I'd even go as far as to say: the setting of instance configuration should rely upon implementation of the proper interface and a setter-method, called when relevant, instead of the constructor parameter.

**Comment by phager-at** [comment]
I've added some more thoughts about general architecture, as I wasn't able to work on it while you implemented your last changes (roughly 2 weeks ago).

TL;DR:
- I still think using interfaces as a form of contract to ensure our implemented features would be good. There'd be solutions for the constructor problem.
- We should think about including the instance configuration (`other`) in the constructor, as it's part of the "configurable instance" feature and should be set via setter defined in the instance_configurable interface.
- Related to the other-setter suggested above, it may be advisable to implement all this in a builder/factory, so the builder/factory can then call all the relevant feature related setters and methods during construction of the rule object

**Comment by daledavies** [comment]
I think I have a solution that satisfies all your feeedback so far. 

- Re-introduce the `rule_interface`
- Move getter and setter for instance config to `instance_configurable` interface
- Remove `rule_base` class entirely
- Use a `rule_trait` trait so rule plugins don't have to implement the basics every time
- Use a factory method to do validation and return a constructed instance object.

Now we hopefully have the best of both worlds, interfaces for everything and rule plugin definition can still be quite simple.

Just tidying up and I'll push as a new commit so you can see the difference.

**Comment by daledavies** [comment]
I'm not too pleased with so many getters and setters, but I guess that's a failing of PHP. Maybe one day we can use property hooks or something!

**Comment by daledavies** [comment]
@phager-at As mentioned above, I've taken your suggestions into account and reworked things in the latest commit. Hopefully this is better, I'm certainly happier with it - thanks for your advise.

Back over to you :)

**Comment by daledavies** [comment]
Rebased to fix conflicts now `main` has been updated.

File: classes/local/rule_instances_controller.php [hunk 8]
```diff
@@ -168,18 +160,61 @@ public function get_rule_instance_records(): array {
     }
 
     /**
-     * Return the rule instance record matching the given instanceid.
+     * Return a list of hydrated rule instance objects.
+     *
+     * @return rule\rule_interface[]|rule\plugin_configurable[]|rule\instance_configurable[]
```
Comment (phager-at) [inline classes/local/rule_instances_controller.php]: My only remark left here: I think `get_rule_instances()`-method's PHPDoc should only mention `rule\rule_interface[]` as return type as it's our base minimum for rule instance classes. As we've also defined in `get_rule_instance_by_id()`.

```suggestion
     * @return rule\rule_interface[]
```

File: classes/local/rule_instances_controller.php [hunk 9]
```diff
@@ -168,18 +160,61 @@ public function get_rule_instance_records(): array {
...
 
-        return $instance;
+        return $instances[$instanceid];
+    }
+
+    /**
+     * Return a list of active rule instance objects, excluding those where either the
+     * rule plugin or instance are disabled, or the rule plugin has not been configured.
+     *
+     * @return array
```
Comment (phager-at) [inline classes/local/rule_instances_controller.php]: (OK seems I lied above, this one can also be more precise. :joy:)

```suggestion
     * @return rule\rule_interface[]
```

**Review by phager-at — APPROVED** [review]
Thanks Dale, for your work and especially your patience here!

I really like the way we're going here, as we're decoupling our higher level plugin code from the implementations in all rule instance plugins with our interfaces. :+1: 

I just found 2 quick remarks about PHPDoc return type definitions, all else LGTM!
(Those are no game-stoppers, so I'll approve the PR now!)

File: classes/local/rule_instances_controller.php (hunk 9 above)
Comment (daledavies) [reply classes/local/rule_instances_controller.php]: Yes these two methods should really return the same thing, only difference is `get_active_rule_instances()` filters out "inactive" instances. 

I.e. those where either the rule plugin or instance are disabled, or the rule plugin has not been configured.

Comment (daledavies) [reply classes/local/rule_instances_controller.php]: Ahh I remember why I used a union type in the docblock for `get_rule_instances()`, frustratingly vscodium isn't able to infer the correct type after the `is_subclass_of()` check in `get_active_rule_instances()`.

Lemme have a think.

Comment (daledavies) [reply classes/local/rule_instances_controller.php]: I guess we may have to accept that vscode can't do this and put up with , at least I can't think of a way to express it in the docblock.

**Comment by daledavies** [comment]
I've updated the docblocks as per your suggestions so will merge now! :)
